File: src/driver.py
"""Command-line driver for the Multi-Target IR Translator.

Module M9.  Owner: Member 1.

Usage
-----
    python -m src.driver --emit=tokens docs/examples/abs.mini
    python -m src.driver --emit=ast    docs/examples/abs.mini
    python -m src.driver --emit=cir    docs/examples/abs.mini
    python -m src.driver --emit=ll     docs/examples/abs.mini
    python -m src.driver --emit=ll     docs/examples/abs.cir   (.cir input)
    python -m src.driver --opt=1 --emit=cir docs/examples/abs.mini
    python -m src.driver --demo-cir                             (hand-built CIR)
    python -m src.driver --demo-stack                           (reg-to-stack)

Stages that are not yet implemented exit with status 3 and a message naming
the module, the owner and the week they are scheduled for, so that the state
of the project is visible from the tool itself rather than only from the plan.
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

from .cir.builder import BuildError, build
from .cir.parser import CirParseError, parse_cir
from .frontend import ast_nodes, parser as mini_parser
from .frontend.lexer import LexError, tokenize
from .frontend.parser import ParseError

EMIT_CHOICES = ["tokens", "ast", "cir", "ll", "wat", "sbc"]


def _demo_cir_module():
    """The abs() module of Figure 2, built by hand.

    It is the fixture for tests/test_cir_printer.py and the golden module
    behind docs/examples/abs.cir.  Now that the builder (M3a) exists,
    --emit=cir on docs/examples/abs.mini produces the same module up to the
    choice of block labels; tests/test_cir_builder.py pins that agreement.
    """
    from .cir.ir import BasicBlock, ConstInt, Function, Instr, Module, Param, Reg, Ty

    x = Reg("x", Ty.I32)
    t0 = Reg("t0", Ty.I1)
    t1 = Reg("t1", Ty.I32)

    entry = BasicBlock("entry", [
        Instr("icmp.slt", Ty.I32, t0, [x, ConstInt(0)]),
        Instr("br.cond", Ty.I1, None, [t0], labels=["then", "exit"]),
    ])
    then = BasicBlock("then", [
        Instr("sub", Ty.I32, t1, [ConstInt(0), x]),
        Instr("ret", Ty.I32, None, [t1]),
    ])
    exit_ = BasicBlock("exit", [
        Instr("ret", Ty.I32, None, [x]),
    ])

    fn = Function("abs", [Param("x", Ty.I32)], Ty.I32, [entry, then, exit_])
    return Module("abs", [], [fn])


def _to_cir(src: str, path: Path, opt_level: int):
    """The CIR module for a source file, optimised to the requested level.

    A ``.cir`` file is read straight through the textual parser rather than
    the front end, which is the contract docs/cir-spec.md section 4 describes:
    a back end can be driven from a checked-in .cir file with no MiniLang
    involved at all.
    """
    from .opt import run as run_opt

    if path.suffix == ".cir":
        module = parse_cir(src, path.stem)
    else:
        module = build(mini_parser.parse(src, str(path)), path.stem)
    return run_opt(module, opt_level)


def _not_yet(stage: str, module: str, owner: str, week: str) -> int:
    print(
        f"error: --emit={stage} is not implemented yet.\n"
        f"       {module} is owned by {owner} and is scheduled for {week}.\n"
        f"       Implemented today: --emit=tokens, --emit=ast, --emit=cir, "
        f"--emit=ll, --demo-cir, --demo-stack.",
        file=sys.stderr,
    )
    return 3


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(
        prog="mtir", description="Multi-Target Intermediate Representation Translator")
    ap.add_argument("source", nargs="?", help="MiniLang source file (.mini)")
    ap.add_argument("--emit", choices=EMIT_CHOICES, default="ast",
                    help="which stage of the pipeline to print")
    ap.add_argument("--opt", type=int, default=0, choices=[0, 1],
                    help="optimisation level: 0 raw, 1 constant folding + "
                         "copy propagation + dead-code elimination")
    ap.add_argument("--run", action="store_true",
                    help="execute the emitted target (Week 9)")
    ap.add_argument("--demo-cir", action="store_true",
                    help="print the hand-built abs() CIR module")
    ap.add_argument("--demo-stack", action="store_true",
                    help="lower the hand-built abs() module to stack form and "
                         "report the naive vs peepholed instruction counts")
    args = ap.parse_args(argv)

    if args.demo_cir:
        from .cir.printer import print_module
        sys.stdout.write(print_module(_demo_cir_module()))
        return 0

    if args.demo_stack:
        from .backend.wasm.reg2stack import count_function, lower_function
        fn = _demo_cir_module().function("abs")
        for label, seq in lower_function(fn).items():
            print(f"{label}:")
            for op in seq:
                print(f"    {op}")
        naive, tuned = count_function(fn)
        saved = 100.0 * (naive - tuned) / naive
        print(f"\nnaive lowering:     {naive} stack instructions")
        print(f"after peephole:     {tuned} stack instructions")
        print(f"reduction:          {saved:.1f}%")
        return 0

    if not args.source:
        ap.error("a source file is required unless --demo-cir or "
                 "--demo-stack is given")

    path = Path(args.source)
    if not path.exists():
        print(f"error: no such file: {path}", file=sys.stderr)
        return 2
    src = path.read_text()

    try:
        if args.emit == "tokens":
            for tok in tokenize(src, str(path)):
                print(f"{tok.line:>4}:{tok.col:<4} {tok.kind.name:<12} {tok.text!r}")
            return 0

        if args.emit == "ast":
            program = mini_parser.parse(src, str(path))
            print(ast_nodes.dump(program))
            return 0

        if args.emit in ("cir", "ll"):
            module = _to_cir(src, path, args.opt)
            if args.emit == "cir":
                from .cir.printer import print_module
                sys.stdout.write(print_module(module))
            else:
                from .backend.llvm.emit_ll import emit as emit_ll
                sys.stdout.write(emit_ll(module))
            return 0

    except LexError as e:
        print(f"{path}:{e}", file=sys.stderr)
        return 1
    except ParseError as e:
        print(f"{path}:{e}", file=sys.stderr)
        return 1
    except (BuildError, CirParseError) as e:
        print(f"{path}:{e}", file=sys.stderr)
        return 1

    pending = {
        "wat": ("WebAssembly back end (M6c)", "Member 4", "Weeks 7-8"),
        "sbc": ("stack bytecode back end (M7)", "Member 4", "Week 9"),
    }
    return _not_yet(args.emit, *pending[args.emit])


if __name__ == "__main__":
    raise SystemExit(main())

File: src/tool/Driver.cpp
// Driver.cpp -- the mtirc command line.  See Driver.h for the exit codes.
//
// Every stage the Python prototype named is now implemented, so there is no
// longer a "not implemented yet" path: an unknown --emit is simply an error.

#include <cstddef>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <utility>
#include <vector>

#include "mtir/tool/Driver.h"

#include "mtir/ast/AST.h"
#include "mtir/backend/llvm/EmitLL.h"
#include "mtir/backend/stackvm/EmitSbc.h"
#include "mtir/backend/stackvm/VM.h"
#include "mtir/backend/wasm/EmitWat.h"
#include "mtir/cir/Builder.h"
#include "mtir/cir/Parser.h"
#include "mtir/frontend/Lexer.h"
#include "mtir/frontend/Parser.h"
#include "mtir/sema/Analyse.h"
#include "mtir/cir/Printer.h"
#include "mtir/cir/Verifier.h"
#include "mtir/opt/Pass.h"

namespace mtir::tool {
namespace {

const char *kUsage = R"(mtirc -- Multi-Target Intermediate Representation Translator

usage: mtirc [options] <file>

  <file>            a .mini source or a .cir module

options:
  --emit=<stage>    tokens | ast | cir | ll | wat | sbc   (default: cir)
  --opt=<level>     0 | 1           (default: 0)
                    1 = constant folding, copy propagation, dead-code
                        elimination, run to a fixed point
  --verify          run the CIR verifier and report any diagnostics
  --run             execute the program on the reference stack VM and print
                    whatever it wrote; the exit status is 4 if it trapped
  --entry=<name>    entry point for --run                 (default: main)
  -o <file>         write to <file> instead of standard output
  --version         print the version and exit
  -h, --help        print this message and exit
)";

struct Options {
  std::string input;
  std::string output;
  std::string emit = "cir";
  int optLevel = 0;
  bool verify = false;
  bool runIt = false;
  std::string entry = "main";
};

bool startsWith(const std::string &s, const std::string &prefix) {
  return s.size() >= prefix.size() && s.compare(0, prefix.size(), prefix) == 0;
}

bool endsWith(const std::string &s, const std::string &suffix) {
  return s.size() >= suffix.size() &&
         s.compare(s.size() - suffix.size(), suffix.size(), suffix) == 0;
}

int unknownStage(const std::string &stage, std::ostream &err) {
  err << "error: unknown stage '" << stage
      << "'. Known stages: tokens, ast, cir, ll, wat, sbc.\n";
  return 1;
}

int fail(const std::string &message, std::ostream &err) {
  err << "error: " << message << "\n";
  return 1;
}

} // namespace

int runDriver(const std::vector<std::string> &args, std::ostream &out,
              std::ostream &err) {
  Options options;
  for (std::size_t i = 0; i < args.size(); ++i) {
    const std::string &arg = args[i];
    if (arg == "-h" || arg == "--help") {
      out << kUsage;
      return 0;
    }
    if (arg == "--version") {
      out << "mtirc 0.1 (C++17)\n";
      return 0;
    }
    if (arg == "--verify") {
      options.verify = true;
    } else if (arg == "--run") {
      options.runIt = true;
    } else if (startsWith(arg, "--entry=")) {
      options.entry = arg.substr(8);
    } else if (startsWith(arg, "--emit=")) {
      options.emit = arg.substr(7);
    } else if (startsWith(arg, "--opt=")) {
      options.optLevel = std::atoi(arg.c_str() + 6);
    } else if (arg == "-o" && i + 1 < args.size()) {
      options.output = args[++i];
    } else if (!arg.empty() && arg[0] == '-') {
      err << "error: unknown option '" << arg << "'\n\n" << kUsage;
      return 1;
    } else if (options.input.empty()) {
      options.input = arg;
    } else {
      return fail("more than one input file given", err);
    }
  }

  if (options.emit != "tokens" && options.emit != "ast" && options.emit != "cir" &&
      options.emit != "ll" && options.emit != "wat" && options.emit != "sbc")
    return unknownStage(options.emit, err);

  if (options.input.empty()) {
    err << kUsage;
    return 1;
  }

  std::ifstream in(options.input, std::ios::binary);
  if (!in) {
    err << "error: no such file: " << options.input << "\n";
    return 2;
  }
  std::ostringstream buffer;
  buffer << in.rdbuf();
  const std::string source = buffer.str();

  // The module takes its name from the file stem, so `mtirc --emit=ll
  // docs/examples/abs.cir` names the module "abs" -- which is what the golden
  // file records.
  std::string moduleName = options.input;
  const std::size_t slash = moduleName.find_last_of("/\\");
  if (slash != std::string::npos)
    moduleName = moduleName.substr(slash + 1);
  const std::size_t dot = moduleName.find_last_of('.');
  if (dot != std::string::npos && dot != 0)
    moduleName = moduleName.substr(0, dot);

  const bool isMiniLang = endsWith(options.input, ".mini");

  // -- the two front-end-only stages ---------------------------------------
  if (options.emit == "tokens" || options.emit == "ast") {
    if (!isMiniLang)
      return fail("--emit=" + options.emit + " needs a .mini source file", err);

    if (options.emit == "tokens") {
      mtir::frontend::LexResult lexed =
          mtir::frontend::tokenize(source, options.input);
      for (const mtir::ast::Token &tok : lexed.tokens) {
        // Formatted with snprintf for the column alignment, then written to
        // the stream rather than to stdout, so a test can read it back.
        char line[512];
        std::snprintf(line, sizeof line, "%4d:%-4d %-12s '%s'\n", tok.loc.line,
                      tok.loc.col,
                      std::string(mtir::ast::tokKindName(tok.kind)).c_str(),
                      tok.text.c_str());
        out << line;
      }
      if (!lexed.ok()) {
        err << mtir::support::format(lexed.diagnostics);
        return 1;
      }
      return 0;
    }

    mtir::frontend::ParseResult parsedAst =
        mtir::frontend::parse(source, options.input);
    if (!parsedAst.ok()) {
      err << mtir::support::format(parsedAst.diagnostics);
      return 1;
    }
    out << mtir::ast::dump(*parsedAst.program);
    return 0;
  }

  // -- .mini or .cir into a CIR module -------------------------------------
  mtir::cir::Module module;
  if (isMiniLang) {
    mtir::frontend::ParseResult parsedAst =
        mtir::frontend::parse(source, options.input);
    if (!parsedAst.ok()) {
      err << mtir::support::format(parsedAst.diagnostics);
      return 1;
    }

    mtir::sema::AnalysisResult analysis =
        mtir::sema::analyse(*parsedAst.program, options.input);
    if (!analysis.ok()) {
      err << mtir::support::format(analysis.diagnostics);
      return 1;
    }

    mtir::cir::BuildResult built =
        mtir::cir::build(*parsedAst.program, *analysis.types, moduleName);
    if (!built.ok()) {
      err << mtir::support::format(built.diagnostics);
      return 1;
    }
    module = std::move(*built.module);
  } else {
    mtir::cir::ParseResult parsed =
        mtir::cir::parseCir(source, moduleName, options.input);
    if (!parsed.ok()) {
      err << mtir::support::format(parsed.diagnostics);
      return 1;
    }
    module = std::move(*parsed.module);
  }

  if (options.optLevel > 0) {
    mtir::opt::PassManager pm = mtir::opt::defaultPipeline(options.optLevel);
    pm.run(module);
  }

  if (options.verify) {
    const auto diags = mtir::cir::verify(module);
    if (!diags.empty()) {
      err << mtir::support::format(diags);
      if (mtir::support::hasErrors(diags))
        return 1;
    }
  }

  // --run short-circuits the emitters: the VM executes the bytecode
  // directly, so nothing needs to be printed as text first.
  if (options.runIt) {
    mtir::backend::stackvm::EmitResult lowered =
        mtir::backend::stackvm::emitModule(module);
    if (!lowered.diagnostics.empty())
      err << mtir::support::format(lowered.diagnostics);
    if (!lowered.ok())
      return 1;

    mtir::backend::stackvm::RunOptions runOptions;
    runOptions.entry = options.entry;
    const mtir::backend::stackvm::RunResult outcome =
        mtir::backend::stackvm::run(lowered.program, runOptions);

    out << outcome.output;
    out.flush();
    if (outcome.trapped) {
      err << "trap: " << outcome.trap << "\n";
      return 4;
    }
    return 0;
  }

  std::string text;
  if (options.emit == "cir") {
    text = mtir::cir::printModule(module);
  } else if (options.emit == "sbc") {
    mtir::backend::stackvm::EmitResult result =
        mtir::backend::stackvm::emitModule(module);
    if (!result.diagnostics.empty())
      err << mtir::support::format(result.diagnostics);
    if (!result.ok())
      return 1;
    text = mtir::backend::stackvm::printProgram(result.program);
  } else if (options.emit == "wat") {
    mtir::backend::wasm::EmitResult result = mtir::backend::wasm::emitModule(module);
    if (!result.diagnostics.empty())
      err << mtir::support::format(result.diagnostics);
    if (!result.ok())
      return 1;
    text = std::move(result.wat);
  } else {
    mtir::backend::llvm::EmitResult result = mtir::backend::llvm::emitModule(module);
    if (!result.ok()) {
      err << mtir::support::format(result.diagnostics);
      return 1;
    }
    text = std::move(result.ir);
  }

  if (options.output.empty()) {
    out << text;
    out.flush();
    return 0;
  }

  std::ofstream file(options.output, std::ios::binary);
  if (!file)
    return fail("cannot write to " + options.output, err);
  file << text;
  return 0;
}

const char *usage() { return kUsage; }

} // namespace mtir::tool
